pkg/cmd/roachtest/run_operation.go

```diff
@@ -0,0 +1,297 @@
+		if opSpec == nil {
+			sleepDuration := time.Duration(baseSleepTime+rng.Intn(baseSleepTime)) * time.Second
+			r.logger.Printf("[%d] couldn't find candidate operation to run, sleeping for %s", workerIdx, sleepDuration)
+			time.Sleep(sleepDuration)
+			continue
+		}
+
+		_ = r.runOperation(ctx, opSpec, rng, workerIdx)
+		func() {
+			r.status.Lock()
+			defer r.status.Unlock()
+			defer r.status.operationRunCompleted.Broadcast()
+			if r.status.lockOperationSelection && opSpec.CanRunConcurrently == registry.OperationCannotRunConcurrently {
+				r.status.lockOperationSelection = false
+			}
+			delete(r.status.running, opSpec.NamePrefix())
+			r.status.lastRun[opSpec.NamePrefix()] = timeutil.Now()
+		}()
+
+		if !runForever {
+			return
+		}
+		// give sometime to worker before picking next operation to run
+		sleepDuration := time.Duration(5+rng.Intn(5)) * time.Minute
+		r.logger.Printf("[%d] going idle for %s", workerIdx, sleepDuration)
+		time.Sleep(sleepDuration)
+	}
+}
+
+// selectOperationToRun picks one operation to run, that hasn't been run for at least waitBeforeNextExecution time.
+func (r *opsRunner) selectOperationToRun(
+	ctx context.Context, rng *rand.Rand, workerID int,
+) *registry.OperationSpec {
+
+	if err := ctx.Err(); err != nil {
+		return nil
+	}
+
+	// randomly select a candidate operation to run
+	opSpec := r.opsToRun[rng.Intn(len(r.opsToRun))]
+
+	r.status.Lock()
+	defer r.status.Unlock()
+
+	// operation which cannotRunConcurrently with other operation is currently running by another worker.
+	// blocking operation selection for now.
+	if r.status.lockOperationSelection {
+		return nil
+	}
+
+	// operation is already running choose another one
+	if _, ok := r.status.running[opSpec.NamePrefix()]; ok {
+		return nil
+	}
+
+	// If the time since the last run of the operation has not exceeded its cadence,
+	// choose another operation
+	if lastRun, ok := r.status.lastRun[opSpec.NamePrefix()]; ok {
+		nextRunTime := lastRun.Add(r.waitBeforeNextExecution)
+		if timeutil.Now().Before(nextRunTime) {
+			return nil
+		}
+	}
+
+	if opSpec.CanRunConcurrently == registry.OperationCannotRunConcurrently {
+		r.status.lockOperationSelection = true
+		// selected operation cannot run concurrently with other operations
+		// wait for other running operation completion
+		for {
+			if len(r.status.running) == 0 {
+				break
+			}
+			r.logger.Printf("[%d] operation: %s waiting for other operation to complete", workerID, opSpec.Name)
+			r.status.operationRunCompleted.Wait()
+		}
+	}
+
+	r.status.running[opSpec.NamePrefix()] = struct{}{}
+	return &opSpec
+}
+
+// runOperation runs a single operation passed in as opSpec parameter within a single operation worker.
+func (r *opsRunner) runOperation(
+	ctx context.Context, opSpec *registry.OperationSpec, rng *rand.Rand, workerIdx int,
+) error {
+	defer func() {
+		if rc := recover(); rc != nil {
+			r.logger.Printf("recovered from panic: %v", rc)
+			debug.PrintStack()
```

Comment:
(Thining out loud), should we emit a datadog event when there's a panic? I think it can be helpful since now we are  ignoring the panic?

Reply:
Make sense, thanks for the suggestion.